ZENKO-3023 phase 1 object squad release process (#27)
scality/citadel: docs/release/disconnecting.md

# Disconecting Releases

From the moment we are in a multi repository environment. We need to disconect the releases of each of our repositories. Define how we release them depending on the role of the repository itself, and then how do we combine them all together.

## Components

> We have multiple libraries, API, tools that we code in house. Let’s treat them as any open source component we would like to use in a project.

Reviewer: Small formatting comment I am mentioning as it helps, me at least, review markdown. I put each sentence on a new line as markdown will merge them all into one paragraph. But for reviewing and detailing it's nice to get that split. Also helps when leaving comments as you can only make line comments and not a word or phrase comments.

Author: True, it doesn't change the output, I'll do that to see how it looks (code wise).

Author: done

> Each component should contain:

Reviewer: I know you didn't want grammar but I would replace should with must, see https://tools.ietf.org/html/rfc2119

Author: HAHAHA fair enough let's change it
```suggestion
Each component must contain:
```

> * Proper code

Reviewer: I would remove this as it ignores documentation repos and config repos ( eslint ), and it is on top people might over prioritize it and skip the others.

Author: Ok
```suggestion
```

* Unit and functional tests to ensure that the component itself work as it should.
* Documentation
* Release notes

> 

Author: I already mentioned test above, do you mind if I modify it to:
```suggestion
* Build chain to build and publish packages
```

> 

Reviewer 2: I'd prefer that you keep "test" on the list.

They will then have their own releases independently from the state of the product itself. They can have 10 or 20 releases before we actually go GA on the project or product side. It doesn’t and shouldn’t matter.

Selecting a component should be as easy and understandable as picking one from the open source world. Either from it’s proper language registry. (bumping the version on the package.json file per example for nodejs projects) or pulling a docker image from a registry. Therefore it is the component responsibility to make those available.

> Each component release is made by the project leader with the help of the CI chapter, advertised internally and externally if needed. Release notes are provided with each release.

Reviewer: I feel project leader should be component leader, or just repository owner(s). Including a CODEOWNERS file per repo could also be another requirement to further distinguish this difference.

Author: So project leader is just the world we are already use to, here at Scality (it's also how the config is named in the bot bert-e but that's just a detail). I don't mind changing the wording to repository owners. As long as it make sense to all of us.

cc: @rahulreddy are you fine with that wording?

The codeowner is also an interesting tool to use. I'm thinking about creating a ticket for it, instead of documenting and tackling it right now, so that there's not too much to do at once, but I could be wrong, what's your thought?  @miniscruff

Reviewer: Code owner gives more control like per folder and maybe per branch, probably not required to add it just yet. But moving forward it may be nice, maybe as a replacement for eve settings it just uses code owners.

Author: So just as an info to ensure that we're all on the same page the settings are for Bert-E, not for eve. Bert-E takes care of gatekeeping the pull requests, and Eve does CI.

But yes, that could be an interesting thing to do, we didn't at first, because we were supporting both Bitbucket and Github and they do not behave the same way, now that we only have to interact with GitHub it makes things easier to implement.

As per the per folder settings, yes I see that this can be very useful especially for docs related stuff where we will need to ensure that William had a look at it per example. But as you're mentioning, let's do baby steps, using `CODEOWNERS` with the whole repo for now, and then we move forward to specific folders. 

We can, therefore add a mention to specify that each repository have a `CODEOWNERS` file up to date. How does that sound?

## Project

Project repository, per example Zenko.

The role of a project repository is to:
* Select *components* with their respective version, either developed in house or open source one.
* Creating the proper build chain to output the desired installation code and packages.
* Integration tests to ensure that all components selected work well together.
* Documentation
* Release notes

Project release should be as small as possible. Per example just bumping the version of a component can unlock feature or bugfixes. Therefore provides value. As long as we keep shipping releases we are unlocking customers by small packs and we get the feedback we need along the way.

> It is the project responsibility to make all required packages for installation or installation instructions available. 

Reviewer: Just to be clear here, you are just saying the documentation is available and not that we have to write it specifically in the project? That is, zenko doesn't need cert-manager docs and can instead link to the component?

Author: Correct, I don't think we need to  re document what's already out there, IMHO. But then in some situation it might be worth writing a word about it.
The idea is mainly to say that the documentation of Zenko resides inside the repo and contain the required information to how to install it and use it. For cert-manager or any other component if we point to the doc of it, it's fine. At the product level though, it could be a different story, but that's not up to the project.

# Disconecting Releases

From the moment we are in a multi repository environment.
We need to disconnect the releases of each of our repositories, define how we release them depending on the role of the repository itself, and how we combine them all together.

## Components

> We have multiple libraries, API, tools that we code in house.

Reviewer 2: Can you add a few examples like vault, cloudserver, arsenal...?

Author: How does this sound?
```suggestion
We have multiple libraries, API, tools that we code in house.
Per example, Arsenal, Cloudserver, Backbeat and so on.
```

Therefore it is the component's responsibility to make those available.

Each component release is made by the project leader with the help of the CI chapter, advertised internally and externally if needed.
Release notes are provided with each release.

## Project

Project repository, per example Zenko.

The role of a project repository is to:
* Select *components* with their respective version, either developed in house or open source one.
* Creating the proper build chain to output the desired installation code and packages.
* Integration tests to ensure that all components selected work well together.
* Documentation
* Release notes

Project release should be as small as possible.
Per example just bumping the version of a component can unlock feature or bugfixes.
Therefore provides value.
As long as we keep shipping releases we are unlocking customers by small packs and we get the feedback we need along the way.

It is the project responsibility to make all required packages for installation or installation instructions available. 

The releases are made by the Project Owner. 

## Product

> The product repository are what is shipped to the customer.

Author: ```suggestion
The product repositories are what is shipped to the customer.
```

The role of a product repository is to:
* Assemble *projects* and *components* with their respective version.
* Proper build chain to output the installation code and packages.
* Integration tests

> * Documentation

Reviewer 2: Maybe a little bit more details.
```suggestion
* Documentation to explain the services provided by the product, how to install, configure, monitor, and solve product issues.
```

# Disconecting Releases

From the moment we are in a multi repository environment.

> We need to disconnect the releases of each of our repositories, define how we release them depending on the role of the repository itself, and how we combine them all together.

Reviewer 3: We must redefine how our products are connected to their component repositories, taking how we intend to release products into consideration.

> We need to disconnect the releases of each of our repositories, define how we release them depending on the role of the repository itself, and how we combine them all together.

Author: ```suggestion
We must redefine how our products are connected to their component repositories, taking how we intend to release products into consideration.
```

> We need to disconect the releases of each of our repositories.
> Define how we release them depending on the role of the repository itself, and then how do we combine them all together.

Reviewer 4: ```suggestion
We need to disconnect the releases of each of our repositories, define how we release them depending on the role of the repository itself, and how we combine them all together.
```

## Components

We have multiple libraries, API, tools that we code in house.

> Per example, Arsenal, Cloudserver, Backbeat and so on.

Reviewer 3: ```suggestion
, for example, Arsenal, CloudServer, and Backbeat.
```

Let’s treat them as any open source component we would like to use in a project.

Each component must contain:
* Unit and functional tests to ensure that the component itself work as it should.

> * Documentation

Author: good point.

* Release notes
* Build chain to build, test, and publish packages

> They will then have their own releases independently from the state of the product itself.

Author: ```suggestion
Each repository will thus have its own release cycle that is independent of the release cycle of the product itself.
```

They can have 10 or 20 releases before we actually go GA on the project or product side.
It doesn’t and shouldn’t matter.

Selecting a component should be as easy and understandable as picking one from the open source world.
Either from it’s proper language registry like bumping the version on the package.json file per example for nodejs projects or pulling a docker image from a registry.

> Therefore it is the component responsibility to make those available.

Reviewer 4: ```suggestion
Therefore it is the component's responsibility to make those available.
```

Each component release is made by the project leader with the help of the CI chapter, advertised internally and externally if needed.
Release notes are provided with each release.

> 

Author: done

## Project

> Project repository, per example Zenko.

Reviewer 3: Fragment. I would move this down or delete it entirely. The next sentence is much more relevant and direct.

> The role of a project repository is to:

Reviewer 3: The purpose of a project repository is to:

> * Select *components* with their respective version, either developed in house or open source one.

Reviewer 3: * Select components (developed in-house or from open source projects) by version.

> * Creating the proper build chain to output the desired installation code and packages.

Reviewer 3: * Create a build chain that outputs the desired installation code and packages.

> * Integration tests to ensure that all components selected work well together.

Reviewer 3: * Trigger integration tests that ensure all selected components interoperate effectively.

> * Documentation to explain the services provided by the project, how to contribute, build, test, configure, run, and monitor the project.

Reviewer 3: * Document the project's services, and how to build, test, configure, run, monitor, and contribute to the project.

> * Release notes

Reviewer 3: * Offer release notes that document changes in the product's function and features.

> Project release should be as small as possible.

Reviewer 3: Project releases should be kept as small as possible. For example, bumping the version of a component might unlock features or bugfixes. If we keep shipping releases, we can advance customers a few at a time and receive needed feedback along the way.

Per example just bumping the version of a component can unlock feature or bugfixes.
Therefore provides value.
As long as we keep shipping releases we are unlocking customers by small packs and we get the feedback we need along the way.

> It is the project responsibility to make all required packages for installation or installation instructions available. 

Reviewer 3: The responsibility for providing instructions and all required packages for installation is at the *project* level, not the *product* level.

> The releases are made by the Project Owner. 

Reviewer 3: Release readiness is determined by the project owner.

## Product

> The product repositories are what is shipped to the customer.

Reviewer 3: Product repositories contain the code and documentation that are shipped to the customer.

> The role of a product repository is to:

Reviewer 3: The purpose of a product repository is to:

> * Assemble *projects* and *components* with their respective version.

Reviewer 3: You need to unpack this (The bad word is "respective".).

* Assemble projects and components under a common _product_ version?

> * Proper build chain to output the installation code and packages.

Reviewer 3: * Create a build chain that outputs the desired installation code and packages.

> * Integration tests

Reviewer 3: * Trigger integration tests 
[that ensure all selected components interoperate effectively.]?

Author: In this case, at the product level it mostly ensure all *projects* (and some components) interoperate effectively

> * Documentation to explain the services provided by the product, how to install, configure, monitor, and solve product issues.

Reviewer 3: * Document the product's services; how to install, configure, run, and monitor the product; and how to solve product issues.

> * Release notes

Reviewer 3: * Offer release notes that document changes in the project's function and features.

> Releases of the product are not defined inside the object squad and they are more based on customer needs.

Reviewer 3: Product releases are driven by customer needs and are not defined within the object squad.

> However it is the object squad responsability to make project releases available for the product and help in the integration work.

Reviewer 3: The object squad, however, is responsible for making project releases available to the product and for helping integrate the projects into the product.
